### PR TITLE
ROX-16908: Add initial Scanner V4 gRPC API services

### PR DESCRIPTION
**Pull request by jvdm**
## Description

🌳 master
├─ :arrow_up_small: **ROX-16908: Add initial Scanner V4 gRPC API services #6659** :point_left: 
├─── :arrow_up_small: ROX-16908: Initialize Scanner backends and services #6663

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Unit test and regression tests added

### N/A

- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

## Testing Performed

TODO

### CODE REVIEW

File: scanner/indexer/service.go [hunk 1]
```diff
@@ -1,12 +1,44 @@
...
+func (s *indexerService) HasIndexReport(ctx context.Context, req *v4.HasIndexReportRequest) (*types.Empty, error) {
+	return nil, status.Errorf(codes.Unimplemented, "method HasIndexReport not implemented")
+}
+
+// RegisterServiceServer registers this service with the given gRPC Server.
+func (s *indexerService) RegisterServiceServer(grpcServer *grpc.Server) {
+	v4.RegisterIndexerServer(grpcServer, s)
+}
+
+// RegisterServiceHandler registers this service with the given gRPC Gateway endpoint.
```
Comment (RTann) [inline scanner/indexer/service.go:40]: Is this needed since we don't use the gateway at this time?

File: scanner/go.mod [hunk 2]
```diff
@@ -3,22 +3,37 @@ module github.com/stackrox/stackrox/scanner/v4
 go 1.20
 
 require (
+	github.com/gogo/protobuf v1.3.2
```
Comment (RTann) [inline scanner/go.mod:6]: This is a deprecated fork of a deprecated package. I think we should try to use `google.golang.org/protobuf` instead. I haven't tried it, but I think you can replace `types.Empty` with this: https://pkg.go.dev/google.golang.org/protobuf@v1.30.0/types/known/emptypb

File: scanner/matcher/service.go [hunk 3]
```diff
@@ -1,12 +1,41 @@
...
-	return &Service{
+func NewMatcherService(matcher *Matcher) (*matcherService, error) {
+	return &matcherService{
 		matcher: matcher,
 	}, nil
 }
+
+func (s *matcherService) GetVulnerabilities(ctx context.Context, req *v4.GetVulnerabilitiesRequest) (*v4.VulnerabilityReport, error) {
+	return nil, status.Errorf(codes.Unimplemented, "method GetVulnerabilities not implemented")
+}
```
Comment (RTann) [inline scanner/matcher/service.go:27]: nit: new line

File: scanner/indexer/service.go (hunk 1 above)
Comment (jvdm) [reply scanner/indexer/service.go:40]: It's part of the interface.

File: scanner/go.mod (hunk 2 above)
Comment (jvdm) [reply scanner/go.mod]: 2eeb86c674..138832e140

File: scanner/matcher/service.go (hunk 3 above)
Comment (jvdm) [reply scanner/matcher/service.go:27]: 2eeb86c674..138832e140

File: scanner/go.mod (hunk 2 above)
Comment (jvdm) [reply scanner/go.mod:6]: The generated protobuf is picking up gogo. If I move to `emptypb` the interface won't match and I won't be able to register my service. This should be dealt with separately.

Comment (RTann) [reply scanner/go.mod:6]: The generated Scanner code is in the main module which requires gogo proto, so I guess we are stuck with it :(

File: scanner/indexer/service.go [hunk 4]
```diff
@@ -1,12 +1,44 @@
...
-	return &Service{
+func NewIndexerService(indexer *Indexer) (*indexerService, error) {
+	return &indexerService{
 		indexer: indexer,
 	}, nil
 }
+
+func (s *indexerService) CreateIndexReport(ctx context.Context, req *v4.CreateIndexReportRequest) (*v4.IndexReport, error) {
+	return nil, status.Errorf(codes.Unimplemented, "method CreateIndexReport not implemented")
+}
```
Comment (RTann) [inline scanner/indexer/service.go:27]: sorry missed this last round. Can you add new line between these functions, too?

File: scanner/indexer/indexer.go [hunk 5]
```diff
@@ -15,7 +17,7 @@ type Indexer struct {
 
 func NewIndexer(ctx context.Context) (*Indexer, error) {
 	// TODO: Update the conn string to something configurable.
-	pool, err := postgres.Connect(ctx, "TODO", "libindex")
+	pool, err := postgres.Connect(ctx, "postgresql:///postgres?host=/var/run/postgresql", "libindex")
```
Comment (RTann) [inline scanner/indexer/indexer.go:20]: curious, do you have something running with this?

**Review by RTann — APPROVED** [review]
Just one nit then 🚢

File: scanner/indexer/indexer.go (hunk 5 above)
Comment (jvdm) [reply scanner/indexer/indexer.go:20]: @RTann This is the default way to access PSQL in the same host (UNIX sockets). Usually, the default database is called "postgres". This is a reasonable guess to run this locally with minimal changes in the person's environment, but your mileage may vary.

File: scanner/indexer/service.go (hunk 4 above)
Comment (jvdm) [reply scanner/indexer/service.go:27]: That's OK, I will follow up in the child PR.

**Pull request merged**